autoscheduler/scraper/management/commands/utils/pdf_parser.py

```diff
@@ -0,0 +1,144 @@
+    if old_pdf_style: # Old pdfs are arranged differently
+        abcdf_slice = slice(4, 9)
+        isuqx_slice = slice(10, 15)
+
+    letter_grades = section_row[abcdf_slice] + section_row[isuqx_slice]
+    letter_grades = {
+        l: int(grade) # Map each letter to the respective amount of grad
+        for l, grade in zip(LETTERS, letter_grades)
+    }
+
+    return letter_grades
+
+
+def parse_page(
+    page_obj: PyPDF2.pdf.PageObject # FIXME I don't like this rule
+) -> List[GradeData]:
+    """ Parses a single page of a PDF, extracting a list of grade data for each section.
+
+        Args:
+            page_obj: A PyPDF2.pdf.PageObject representing the current page
+        Returns:
+            A list of GradeData objects, with the GPA as None
+    """
+
+    text = sanitize_page(page_obj) # Splits the text into separate lines
+    i = 0 # The current row in the pdf we're on
+    grade_data = [] # list of GradeData objects
+
+    while i < len(text):
+        skip_count = get_pdf_skip_count(text[i])
+        old_pdf_style = skip_count[0]
+        count = skip_count[1] # The actual count of how many rows should be skipped
+
+        if count != -1:
+            i += count
+        else: # These are the rows that have actual grade data on them
+            # Split the row into separate sections
+            section_row = text[i : i + LEN_SECTION_ROW]
+            try:
+                dept, course_num, section_num = section_row[0].split("-")
+
+                letter_grades = extract_letter_grades(section_row, old_pdf_style)
+
+                grade = GradeData(dept, course_num, section_num, letter_grades, None)
+
+                grade_data.append(grade)
+
+                i += LEN_SECTION_ROW
+            except ValueError: # When would this happen? What causes it?
+                i += LEN_SECTION_ROW - 1
+
+    return grade_data
+
+def calculate_gpa(letter_grades: Dict) -> float:
+    """ Given a series of letter grades, calculates the GPA of the section.
+
+        Args:
+            letter_grades: A list of integers representing how many students got
+                           each letter grade
+        Returns:
+            The calculated gpa.
+    """
+    weights = [4.0, 3.0, 2.0, 1.0, 0.0] # A to F GPAs, respectively
+    grades = [letter_grades[char] for char in ["A", "B", "C", "D", "F"]]
+    num_students = sum(grades)
+
+    gpa = 0.0
+    for students_with_grade, weight in zip(grades, weights):
+        gpa += students_with_grade * weight
+    return gpa / num_students
+
+def parse_pdf(pdf_path: str) -> List[GradeData]:
+    """ Loads the pdf from the given path and calls parse_page on each page, then collects
+        the results into one array.
+
+        Args:
+            pdf_path: The path to the pdf
+        Returns:
+            A list of GradeData, with each GPA calculated
+    """
+
+    with open(pdf_path, "rb") as file:
+        pdf_reader = PyPDF2.PdfFileReader(file)
+        pdf_data = []

```

Comment:
Since this is never modified anywhere, why does this function return an empty array? Or am I missing something

Reply:
Yeah it should be `pdf_data += page_data` after the `for grade in page_data:` loop